pkg/model/diamond.go

```diff
@@ -0,0 +1,265 @@
+	return string(s)
+}
+
+// DiamondDescriptor models a diamond's metadata
+type DiamondDescriptor struct {
+	DiamondID       string            `json:"diamondID" yaml:"diamondID"`
+	StartTime       time.Time         `json:"startTime" yaml:"startTime"` // when the diamond was initialized
+	EndTime         time.Time         `json:"endTime,omitempty" yaml:"endTime,omitempty"`
+	StateChangeTime time.Time         `json:"stateChangeTime" yaml:"stateChangeTime"` // last state change
+	State           DiamondState      `json:"state" yaml:"state"`
+	Mode            ConflictMode      `json:"mode" yaml:"mode"`
+	HasConflicts    bool              `json:"hasConflicts,omitempty" yaml:"hasConflicts,omitempty"`
+	HasCheckpoints  bool              `json:"hasCheckpoints,omitempty" yaml:"hasCheckpoints,omitempty"`
+	Tag             string            `json:"tag,omitempty" yaml:"tag,omitempty"`
+	BundleID        string            `json:"bundleID,omitempty" yaml:"bundleID,omitempty"`
+	Splits          []SplitDescriptor `json:"splits,omitempty" yaml:"splits,omitempty"`
+	_               struct{}
+}
+
+// DiamondDescriptors is a sortable slice of DiamondDescriptor
+type DiamondDescriptors []DiamondDescriptor
+
+func (b DiamondDescriptors) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+func (b DiamondDescriptors) Len() int {
+	return len(b)
+}
+func (b DiamondDescriptors) Less(i, j int) bool {
+	if !b[i].StartTime.IsZero() && !b[j].StartTime.IsZero() {
+		return b[i].StartTime.Before(b[j].StartTime)
+	}
+	return b[i].DiamondID < b[j].DiamondID
+}
+
+// Last bundle descriptor in slice
+func (b DiamondDescriptors) Last() DiamondDescriptor {
+	return b[len(b)-1]
+}
+
+func defaultDiamondDescriptor() *DiamondDescriptor {
+	ksuid, err := ksuid.NewRandom()
+	if err != nil {
+		panic(fmt.Sprintf("cannot generate random ksuid: %v", err))
+	}
+	return &DiamondDescriptor{
+		DiamondID: ksuid.String(),
+		StartTime: GetBundleTimeStamp(),
+		State:     DiamondInitialized,
+		Mode:      EnableConflicts,
+	}
+}
+
+// NewDiamondDescriptor builds a new DiamondDescriptor
+func NewDiamondDescriptor(opts ...DiamondDescriptorOption) *DiamondDescriptor {
+	d := defaultDiamondDescriptor()
+	for _, apply := range opts {
+		apply(d)
+	}
+	return d
+}
+
+// SplitDescriptor models the metadata about a given split within a diamond
+type SplitDescriptor struct {
+	SplitID               string        `json:"splitID" yaml:"splitID"`
+	StartTime             time.Time     `json:"startTime" yaml:"startTime"`
+	EndTime               time.Time     `json:"endTime,omitempty" yaml:"endTime,omitempty"`
+	State                 SplitState    `json:"state" yaml:"state"`
+	Contributors          []Contributor `json:"contributors" yaml:"contributors"` // Contributor for the bundle
+	SplitEntriesFileCount uint64        `json:"count" yaml:"count"`               // Number of index files in this bundle
+	Tag                   string        `json:"tag,omitempty" yaml:"tag,omitempty"`
```

Comment:
What is tag?

Reply:
same intent as explained above

> Tag is just an informative string to be kept in the logs and in list commands.
> I am basically using this to triage log from many splits and coordinator pods

Reply:
note to self that this is in the "nice to have" [section at the end](https://github.com/oneconcern/datamon/blob/diamond-metadata-paths/docs/proposals/diamond-design.md#nice-to-have--future) of the design document